File: java/Samples/Biometrics/VB/SimpleFacesSampleVB/GeneralizeFaces.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images

Partial Public Class GeneralizeFaces
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _subject As NSubject = Nothing

#End Region

#Region "Public constructor"

	Private privateBiometricClient As NBiometricClient
	Public Property BiometricClient() As NBiometricClient
		Get
			Return privateBiometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			privateBiometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub OnCreateTemplateCompleted(ByVal result As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnCreateTemplateCompleted), result)
		Else
			Try
				Dim status As NBiometricStatus = BiometricClient.EndCreateTemplate(result)
				If status = NBiometricStatus.Ok Then
					lblStatus.Text = "Status: Ok"
					lblStatus.ForeColor = Color.Green
					btnSaveTemplate.Enabled = True
					faceView.Face = _subject.Faces.Last()
				Else
					lblStatus.Text = String.Format("Status: {0}", status)
					lblStatus.ForeColor = Color.Red
				End If
			Catch ex As Exception
				lblStatus.Text = "Status: error occurred"
				lblStatus.ForeColor = Color.Red
				Utils.ShowException(ex)
			End Try
			btnOpenImages.Enabled = True
		End If
	End Sub

	Private Sub BtnOpenImagesClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenImages.Click
		_subject = Nothing
		faceView.Face = Nothing
		btnSaveTemplate.Enabled = False
		lblImageCount.Text = "0"
		lblStatus.Visible = False

		If OpenFileDialog.ShowDialog() = DialogResult.OK Then
			_subject = New NSubject()
			For Each fileName In openFileDialog.FileNames
				Dim face As New NFace() With {.FileName = fileName, .SessionId = 1}
				_subject.Faces.Add(face)
			Next fileName
			lblImageCount.Text = OpenFileDialog.FileNames.Length.ToString()
			lblStatus.Text = "Status: performing extraction and generalization"
			lblStatus.Visible = True
			lblStatus.ForeColor = Color.Black
			BiometricClient.BeginCreateTemplate(_subject, AddressOf OnCreateTemplateCompleted, Nothing)
			btnOpenImages.Enabled = False
		End If
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If _subject IsNot Nothing Then
			If saveTemplateDialog.ShowDialog() = DialogResult.OK Then
				Try
					Using templateBuffer = _subject.GetTemplateBuffer()
						File.WriteAllBytes(saveTemplateDialog.FileName, templateBuffer.ToArray())
					End Using
				Catch ex As Exception
					Utils.ShowException(ex)
				End Try
			End If
		End If
	End Sub

	Private Sub GeneralizeFacesLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			OpenFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		End If
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/SimpleFacesSampleCS/GeneralizeFaces.cs
using System;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class GeneralizeFaces : UserControl
	{
		#region Public constructor

		public GeneralizeFaces()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NSubject _subject = null;

		#endregion

		#region Public constructor

		public NBiometricClient BiometricClient { get; set; }

		#endregion

		#region Private methods

		private void OnCreateTemplateCompleted(IAsyncResult result)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnCreateTemplateCompleted), result);
			}
			else
			{
				try
				{
					NBiometricStatus status = BiometricClient.EndCreateTemplate(result);
					if (status == NBiometricStatus.Ok)
					{
						lblStatus.Text = "Status: Ok";
						lblStatus.ForeColor = Color.Green;
						btnSaveTemplate.Enabled = true;
						faceView.Face = _subject.Faces.Last();
					}
					else
					{
						lblStatus.Text = string.Format("Status: {0}", status);
						lblStatus.ForeColor = Color.Red;
					}
				}
				catch (Exception ex)
				{
					lblStatus.Text = "Status: error occurred";
					lblStatus.ForeColor = Color.Red;
					Utils.ShowException(ex);
				}
				btnOpenImages.Enabled = true;
			}
		}

		private void BtnOpenImagesClick(object sender, EventArgs e)
		{
			_subject = null;
			faceView.Face = null;
			btnSaveTemplate.Enabled = false;
			lblImageCount.Text = "0";
			lblStatus.Visible = false;

			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				_subject = new NSubject();
				foreach (var name in openFileDialog.FileNames)
				{
					NFace face = new NFace()
					{
						FileName = name,
						SessionId = 1 // all faces with same session will be generalized
					};
					_subject.Faces.Add(face);
				}
				lblImageCount.Text = openFileDialog.FileNames.Length.ToString();
				lblStatus.Text = "Status: performing extraction and generalization";
				lblStatus.Visible = true;
				lblStatus.ForeColor = Color.Black;
				BiometricClient.BeginCreateTemplate(_subject, OnCreateTemplateCompleted, null);
				btnOpenImages.Enabled = false;
			}
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (_subject != null)
			{
				if (saveTemplateDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						using (var templateBuffer = _subject.GetTemplateBuffer())
						{
							File.WriteAllBytes(saveTemplateDialog.FileName, templateBuffer.ToArray());
						}
					}
					catch (Exception ex)
					{
						Utils.ShowException(ex);
					}
				}
			}
		}

		private void GeneralizeFacesLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				openFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);
			}
		}

		#endregion
	}
}
